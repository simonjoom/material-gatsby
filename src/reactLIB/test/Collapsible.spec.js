import React from 'react';
import { shallow, mount } from 'enzyme';
import Collapsible from '../src/Collapsible';
import CollapsibleItem from '../src/CollapsibleItem';
import mocker from './helper/new-mocker';

describe('<Collapsible />', () => {
  let wrapper;
  const collapsibleInitMock = jest.fn();
  const collapsibleInstanceDestroyMock = jest.fn();
  const collapsibleMock = {
    init: (el, options) => {
      collapsibleInitMock(options);
      return {
        destroy: collapsibleInstanceDestroyMock
      };
    }
  };
  const restore = mocker('Collapsible', collapsibleMock);

  const options = {
    onOpenStart: () => {},
    inDuration: 100,
    outDuration: 50
  };

  beforeEach(() => {
    collapsibleInitMock.mockClear();
  });

  afterAll(() => {
    restore();
  });

  test('initializes Collapsible with options', () => {
    wrapper = shallow(<Collapsible options={options} />);
    expect(collapsibleInitMock).toHaveBeenCalledWith(options);
  });

  test('Destroyed when unmounted', () => {
    wrapper = shallow(<Collapsible options={options} />);
    wrapper.unmount();
    expect(collapsibleInstanceDestroyMock).toHaveBeenCalled();
  });

  test('renders', () => {
    wrapper = shallow(<Collapsible />);
    expect(wrapper).toMatchSnapshot();
  });

  test('with accordion', () => {
    wrapper = shallow(<Collapsible accordion />);
    expect(wrapper).toMatchSnapshot();
  });

  test('with invalid elements', () => {
    wrapper = shallow(
      <Collapsible>
        {null}
        <CollapsibleItem header="First">A</CollapsibleItem>
        {null}
        <CollapsibleItem header="Second">B</CollapsibleItem>
      </Collapsible>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.children().length).toEqual(2);
  });

  test('with expanded prop', () => {
    wrapper = mount(
      <Collapsible>
        <CollapsibleItem expanded header="First">
          A
        </CollapsibleItem>
        <CollapsibleItem header="Second">B</CollapsibleItem>
      </Collapsible>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('updates activeKey', () => {
    wrapper = mount(
      <Collapsible accordion>
        <CollapsibleItem header="A">A</CollapsibleItem>
        <CollapsibleItem header="B">B</CollapsibleItem>
        <CollapsibleItem header="C">C</CollapsibleItem>
        <CollapsibleItem header="D">D</CollapsibleItem>
        <CollapsibleItem header="E">E</CollapsibleItem>
      </Collapsible>
    );

    expect(wrapper.state().activeKey).toBeUndefined();
    expect(wrapper).toMatchSnapshot();
    wrapper
      .find('.collapsible-header')
      .at(3)
      .simulate('click');
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state().activeKey).toEqual(3);
  });

  test('defaultActiveKey', () => {
    const activeKey = 0;
    wrapper = mount(
      <Collapsible accordion defaultActiveKey={activeKey}>
        <CollapsibleItem header="First">A</CollapsibleItem>
        <CollapsibleItem header="Second">B</CollapsibleItem>
      </Collapsible>
    );

    wrapper
      .find('.collapsible-header')
      .at(0)
      .simulate('click');
    expect(wrapper.state().activeKey).toBeNull();
    expect(wrapper).toMatchSnapshot();
    wrapper
      .find('.collapsible-header')
      .at(1)
      .simulate('click');
    expect(wrapper.state().activeKey).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });

  test('onSelect with right called with right argument', () => {
    const onSelect = jest.fn();
    wrapper = mount(
      <Collapsible accordion onSelect={onSelect}>
        <CollapsibleItem header="A">A</CollapsibleItem>
        <CollapsibleItem header="B">B</CollapsibleItem>
        <CollapsibleItem header="C">C</CollapsibleItem>
        <CollapsibleItem header="D">D</CollapsibleItem>
        <CollapsibleItem header="E">E</CollapsibleItem>
      </Collapsible>
    );

    const clickedIndex = 3;
    wrapper
      .find('.collapsible-header')
      .at(clickedIndex)
      .simulate('click');
    expect(onSelect).toHaveBeenCalledWith(clickedIndex);
  });
});
