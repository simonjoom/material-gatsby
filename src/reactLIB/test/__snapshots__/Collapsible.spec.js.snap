// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapsible /> defaultActiveKey 1`] = `
<Collapsible
  accordion={true}
  defaultActiveKey={0}
>
  <ul
    className="collapsible"
    data-collapsible="accordion"
  >
    <CollapsibleItem
      eventKey={0}
      expanded={false}
      header="First"
      key=".0"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          First
        </div>
        <div
          className="collapsible-body"
        >
          A
        </div>
      </li>
    </CollapsibleItem>
    <CollapsibleItem
      eventKey={1}
      expanded={false}
      header="Second"
      key=".1"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          Second
        </div>
        <div
          className="collapsible-body"
        >
          B
        </div>
      </li>
    </CollapsibleItem>
  </ul>
</Collapsible>
`;

exports[`<Collapsible /> defaultActiveKey 2`] = `
<Collapsible
  accordion={true}
  defaultActiveKey={0}
>
  <ul
    className="collapsible"
    data-collapsible="accordion"
  >
    <CollapsibleItem
      eventKey={0}
      expanded={false}
      header="First"
      key=".0"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          First
        </div>
        <div
          className="collapsible-body"
        >
          A
        </div>
      </li>
    </CollapsibleItem>
    <CollapsibleItem
      eventKey={1}
      expanded={true}
      header="Second"
      key=".1"
      node="div"
      onSelect={[Function]}
    >
      <li
        className="active"
      >
        <div
          className="collapsible-header active"
          onClick={[Function]}
        >
          Second
        </div>
        <div
          className="collapsible-body"
        >
          B
        </div>
      </li>
    </CollapsibleItem>
  </ul>
</Collapsible>
`;

exports[`<Collapsible /> renders 1`] = `
<ul
  className="collapsible"
  data-collapsible="expandable"
/>
`;

exports[`<Collapsible /> updates activeKey 1`] = `
<Collapsible
  accordion={true}
>
  <ul
    className="collapsible"
    data-collapsible="accordion"
  >
    <CollapsibleItem
      eventKey={0}
      expanded={false}
      header="A"
      key=".0"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          A
        </div>
        <div
          className="collapsible-body"
        >
          A
        </div>
      </li>
    </CollapsibleItem>
    <CollapsibleItem
      eventKey={1}
      expanded={false}
      header="B"
      key=".1"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          B
        </div>
        <div
          className="collapsible-body"
        >
          B
        </div>
      </li>
    </CollapsibleItem>
    <CollapsibleItem
      eventKey={2}
      expanded={false}
      header="C"
      key=".2"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          C
        </div>
        <div
          className="collapsible-body"
        >
          C
        </div>
      </li>
    </CollapsibleItem>
    <CollapsibleItem
      eventKey={3}
      expanded={false}
      header="D"
      key=".3"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          D
        </div>
        <div
          className="collapsible-body"
        >
          D
        </div>
      </li>
    </CollapsibleItem>
    <CollapsibleItem
      eventKey={4}
      expanded={false}
      header="E"
      key=".4"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          E
        </div>
        <div
          className="collapsible-body"
        >
          E
        </div>
      </li>
    </CollapsibleItem>
  </ul>
</Collapsible>
`;

exports[`<Collapsible /> updates activeKey 2`] = `
<Collapsible
  accordion={true}
>
  <ul
    className="collapsible"
    data-collapsible="accordion"
  >
    <CollapsibleItem
      eventKey={0}
      expanded={false}
      header="A"
      key=".0"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          A
        </div>
        <div
          className="collapsible-body"
        >
          A
        </div>
      </li>
    </CollapsibleItem>
    <CollapsibleItem
      eventKey={1}
      expanded={false}
      header="B"
      key=".1"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          B
        </div>
        <div
          className="collapsible-body"
        >
          B
        </div>
      </li>
    </CollapsibleItem>
    <CollapsibleItem
      eventKey={2}
      expanded={false}
      header="C"
      key=".2"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          C
        </div>
        <div
          className="collapsible-body"
        >
          C
        </div>
      </li>
    </CollapsibleItem>
    <CollapsibleItem
      eventKey={3}
      expanded={true}
      header="D"
      key=".3"
      node="div"
      onSelect={[Function]}
    >
      <li
        className="active"
      >
        <div
          className="collapsible-header active"
          onClick={[Function]}
        >
          D
        </div>
        <div
          className="collapsible-body"
        >
          D
        </div>
      </li>
    </CollapsibleItem>
    <CollapsibleItem
      eventKey={4}
      expanded={false}
      header="E"
      key=".4"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          E
        </div>
        <div
          className="collapsible-body"
        >
          E
        </div>
      </li>
    </CollapsibleItem>
  </ul>
</Collapsible>
`;

exports[`<Collapsible /> with accordion 1`] = `
<ul
  className="collapsible"
  data-collapsible="accordion"
/>
`;

exports[`<Collapsible /> with expanded prop 1`] = `
<Collapsible
  accordion={false}
>
  <ul
    className="collapsible"
    data-collapsible="expandable"
  >
    <CollapsibleItem
      eventKey={0}
      expanded={true}
      header="First"
      key=".0"
      node="div"
      onSelect={[Function]}
    >
      <li
        className="active"
      >
        <div
          className="collapsible-header active"
          onClick={[Function]}
        >
          First
        </div>
        <div
          className="collapsible-body"
        >
          A
        </div>
      </li>
    </CollapsibleItem>
    <CollapsibleItem
      eventKey={1}
      expanded={false}
      header="Second"
      key=".1"
      node="div"
      onSelect={[Function]}
    >
      <li
        className=""
      >
        <div
          className="collapsible-header"
          onClick={[Function]}
        >
          Second
        </div>
        <div
          className="collapsible-body"
        >
          B
        </div>
      </li>
    </CollapsibleItem>
  </ul>
</Collapsible>
`;

exports[`<Collapsible /> with invalid elements 1`] = `
<ul
  className="collapsible"
  data-collapsible="expandable"
>
  <CollapsibleItem
    eventKey={1}
    expanded={false}
    header="First"
    key=".1"
    node="div"
    onSelect={[Function]}
  >
    A
  </CollapsibleItem>
  <CollapsibleItem
    eventKey={3}
    expanded={false}
    header="Second"
    key=".3"
    node="div"
    onSelect={[Function]}
  >
    B
  </CollapsibleItem>
</ul>
`;
